Incorporating consideration of parallel 2WTs (#136)
powsybl/powsybl-optimizer: docs/optimizer/acOptimalPowerflow.md

- The reactive power $\boldsymbol{Q_{i,vsc}}$ produced by voltage source converter station $vsc$ is included in $[\min(qP_{vsc}, qp_{vsc}, qp_{vsc}^0)$; $\max(QP_{vsc}, Qp_{vsc}, Qp_{vsc}^0)]$.
**The bounds are therefore rectangular, not trapezoidal.**

## Parallel transformers

Transformers connected in parallel (sharing the same pair of buses, or forming a closed loop of transformers inside a single substation) should keep the same transformation ratio: letting them diverge would create circulating reactive flows between the parallel branches. The optimizer therefore detects such groups (called *bundles*) automatically, and constrains each bundle to a single shared ratio.

The quantity that must be equalized is the *effective* per-unit ratio $c_{ij} \boldsymbol{\rho_{ij}}$, where $c_{ij}$ is the constant (off-tap) per-unit ratio of the transformer (the "cst ratio (pu)" column of the network data): this is the ratio entering the flow equations, whose mismatch drives circulating flows. When all members of a bundle are identical units ($c_{ij}$ equal), this is equivalent to equalizing the tap ratios themselves.

> The effective ratio applies from terminal 1 to terminal 2 *as declared* in the network, and nothing in the data model forces two physically parallel transformers to be declared in the same direction. Each member is therefore qualified against a canonical direction of its bundle (that of its first member in id order): a member declared in the same direction is *direct* (orientation +1), a member declared in the opposite direction is *reversed* (orientation -1). The comparison uses the nominal-voltage side of terminal 1, falling back to the terminal-1 bus for equal-nominal-voltage bundles sharing a single bus pair; an equal-nominal-voltage bundle coming from a loop cannot be oriented and is released with a warning instead of being tied blindly.

Reviewer: ```suggestion
The effective ratio applies from terminal 1 to terminal 2 *as declared* in the network, and nothing in the data model forces two physically parallel transformers to be declared in the same direction. Each member is therefore qualified against a canonical direction of its bundle (that of its first member in id order): a member declared in the same direction is *direct* (orientation +1), a member declared in the opposite direction is *reversed* (orientation -1). The comparison uses the nominal voltage side of terminal 1, falling back to the terminal 1 bus for equal nominal voltage bundles sharing a single bus pair; an equal nominal voltage bundle that comes from a loop cannot be oriented and is relaxed with a warning instead of being tied blindly.
```

Author: Done in the next commit.

- **empty interval** ($\rho_B^{min} > \rho_B^{max}$): the members' effective ranges are disjoint, which signals inconsistent input data (e.g. disjoint tap ranges, or rated voltages making a common effective ratio unreachable). The optimizer falls back to a best-effort behavior, fixing each member as close as possible to the center of the gap $\frac{\rho_B^{min} + \rho_B^{max}}{2}$ (its inverse for a reversed member), clamped to its own declared effective range $[c_{ij} \rho_{ij}^{min}, c_{ij} \rho_{ij}^{max}]$.

A bundle is tied through $(7)$ only if **all** its members are optimized variable-ratio transformers (specified in `param_transformers.txt`, see [Configuration of the run](inputs.md#configuration-of-the-run)). A member that is not optimized is frozen at its current tap, which pins the shared ratio to that value and collapses the bundle to the single-point or empty case above. As a consequence, declaring only a subset of a parallel group as variable does not optimize that subset freely: those transformers are fixed at the common ratio, again to avoid circulating flows.

> This grouping interacts with the second optimization after tap rounding (see [Solving](#solving)): constraint $(7)$ is released before rounding, so each transformer is rounded to the nearest tap of its own table independently. Members of a bundle may therefore end up on different discrete taps if their tap tables differ; the shared ratio is guaranteed only for the continuous solving.

Reviewer: ```suggestion
This grouping interacts with the second optimization after tap rounding (see [Solving](#solving)): constraint $(7)$ is relaxed before rounding, so each transformer is rounded to the nearest tap of its own table independently. Members of a bundle may therefore end up on different discrete taps if their tap tables differ; the shared ratio is guaranteed only for the continuous solving.
```

Author: Done in the next commit.

> Finally, if at solve time a member of a tied bundle turns out not to be an optimized variable-ratio branch — for instance because it lies outside the main connex component, or has a near-zero impedance — the whole bundle is released: its members are optimized independently for that run, and the event is logged.

Reviewer: ```suggestion
Finally, if at solve time a member of a tied bundle turns out not to be an optimized variable-ratio branch — for instance because it lies outside the main connex component, or has a near-zero impedance — the whole bundle is relaxed: its members are optimized independently for that run, and the event is logged.
```

Author: Done in the next commit.
